File: src/translated/actions.cpp
//
//  actions.cpp
//
//  Created by Eliana Fernandes
//  Copyright (c) 2015 Eliana Fernandes. All rights reserved.
//

#include "actions.h"

int AddSolutionComponent(ind indices, data dados, ant& ants, double& temp_cost){
    // add solution component
    ants.x[indices.node][indices.prod]=indices.hub;
#ifdef HISTORY
	ofstream myfile;
	myfile.open ("history1.txt",ios::app);
	myfile<<endl<<"Solution Component Added"<<endl;
	myfile << "ants.x["<<indices.node<<"]["<<indices.prod<<"]= " << indices.hub << endl;
	myfile.close();
#endif
    // remove solution component of available solutions
    ants.avail_tau[indices.node][indices.hub][indices.prod]=0;
    // Compute transfer and collection costs
    // update costs
#ifdef HISTORY
	myfile.open ("history1.txt",ios::app);
	myfile<<endl<<"update transfer and collection costs costs from: "<<endl;
	myfile << "ants.cost "<<ants.cost << endl;
	myfile << "temp_cost "<<temp_cost << endl;
	myfile.close();
#endif
    temp_cost = ants.cost;
    ants.cost=temp_cost+dados.d[indices.node][indices.hub]*(dados.chi[indices.prod]*dados.O[indices.node][indices.prod]+dados.delta[indices.prod]*dados.D[indices.node][indices.prod]);
    temp_cost = ants.cost;
#ifdef HISTORY
	myfile.open ("history1.txt",ios::app);
	myfile<<endl<<"update transfer and collection costs costs to: "<<endl;
	myfile << "ants.cost "<<ants.cost << endl;
	myfile << "temp_cost "<<temp_cost << endl;
	myfile.close();
#endif
	// Compute fixed costs if the node is to be a hub
	if(indices.node==indices.hub){
#ifdef HISTORY
		myfile.open ("history1.txt",ios::app);
		myfile<<endl<<"update fixed costs from: "<<endl;
		myfile << "ants.cost "<<ants.cost << endl;
		myfile << "temp_cost "<<temp_cost << endl;
		myfile.close();
#endif
		temp_cost = ants.cost;
		ants.cost=temp_cost+dados.f[indices.hub][indices.prod];
		temp_cost = ants.cost;
#ifdef HISTORY
		myfile.open ("history1.txt",ios::app);
		myfile<<endl<<"update fixed costs to: "<<endl;
		myfile << "ants.cost "<<ants.cost << endl;
		myfile << "temp_cost "<<temp_cost << endl;
		myfile.close();
#endif
	}
    return 0;   
}

int LocalPheromoneUpdate(int prod,int hub, int node, aco& a_param){
    double temp_tau;
#ifdef TAU_HIST
	ofstream taufile;
	taufile.open ("tau_hist.txt",ios::app);
	taufile << "tau local updated"<< endl;
	taufile << "a.tau"<<"["<<node<<"]["<<hub<<"]["<<prod<<"] = "<< a_param.tau[hub][node][prod] << endl;
	taufile.close();
#endif
    temp_tau=a_param.tau[node][hub][prod];
    a_param.tau[node][hub][prod]=(1-RHO)*temp_tau+RHO*a_param.tau0[node][hub][prod];
#ifdef TAU_HIST
	taufile.open ("tau_hist.txt",ios::app);
	taufile << "a.tau"<<"["<<node<<"]["<<hub<<"]["<<prod<<"] = "<< a_param.tau[hub][node][prod] << endl;
	taufile.close();
#endif
    return 0;
}

int ApplySingleAllocationRules(int prod, int hub, int node, int nr_nodes, ant& ants, int is_node, int is_not_hub){
    int i,j;
#ifdef HISTORY
	ofstream myfile;
	myfile.open ("history1.txt",ios::app);
	myfile<<endl<<"SOLUTION COMPONENTS MADE UNAVAILABLE SA RULES"<<endl;
	myfile.close();
#endif
    // node SA
    // if node i is assigned to hub j, it cannot be assigned to another hub (Single Allocation)
    // prevent multiple allocation of node i for product p
	if(is_node>0)
		for(j=0;j<nr_nodes;j++){
			if(ants.avail_tau[node][j][prod]>0){
				ants.avail_tau[node][j][prod]=0;
#ifdef HISTORY
				myfile.open ("history1.txt",ios::app);
				myfile << "ants.avail_tau"<<"["<<node<<"]["<<j<<"]["<<prod<<"] = "<< ants.avail_tau[node][j][prod] << endl;
				myfile.close();
#endif
			}
		}
    // node_not_hub SA
    // once a node is allocated to a hub, it cannot become a hub because of single allocation
    // (the hubs are alocated to themselves)
    // so all connections to that node are forbidden
    // if that node is a hub, this cannot be done because hubs don't have connection limitations
		if(is_not_hub>0)
			for(i=0;i<nr_nodes;i++){
				if(ants.avail_tau[i][node][prod]>0){
					ants.avail_tau[i][node][prod]=0;
#ifdef HISTORY
					myfile.open ("history1.txt",ios::app);
					myfile << "ants.avail_tau"<<"["<<i<<"]["<<node<<"]["<<prod<<"] = "<< ants.avail_tau[i][node][prod] << endl;
					myfile.close();
#endif
				}
			}
    return 0;
}

int ApplyLkRules(int prod, int hub, data dados,ant& ants){
    
    if(dados.L[prod]<dados.nbProducts){
#ifdef HISTORY
		ofstream myfile;
		myfile.open ("history1.txt",ios::app);
		myfile<<endl<<"SOLUTION COMPONENTS MADE UNAVAILABLE Lk Rules"<<endl;
		myfile.close();
#endif
        
        int i,p;        
        hubpcount np;        
        np = checkProdConnects(dados.nbProducts, dados.nbNodes, ants);
        
        int iff=0;
        // if the hub is "full" (reached full product capacity)
        if(np.prods[hub]>=dados.L[hub]){
            iff=1;
            for(p=0;p<dados.nbProducts;p++)
				if(ants.x[hub][p]!=hub) //Se já for hub nao ha problema
                for(i=0;i<dados.nbNodes;i++){
					if(ants.avail_tau[i][hub][p]>0){
						ants.avail_tau[i][hub][p]=0;
#ifdef HISTORY
						myfile.open ("history1.txt",ios::app);
						myfile << "ants.avail_tau"<<"["<<i<<"]["<<hub<<"]["<<p<<"] = "<< ants.avail_tau[i][hub][p] << endl;
						myfile.close();
#endif
					}
				}
		}
		delete [] np.prods;
	}	
    return 0;
}


int UpdateAvailableCapacities(int prod, int hub, int node, data dados, ant& ants, int k){
#ifdef HISTORY
	ofstream myfile;
	myfile.open ("history1.txt",ios::app);
	myfile<<endl<<"CAPACITY UPDATE FROM "<<ants.avail_cap[hub][prod];
	myfile.close();
#endif
	double temp_cap;
	temp_cap=ants.avail_cap[hub][prod]; //NEW
    ants.avail_cap[hub][prod]=temp_cap-dados.O[node][prod];
#ifdef HISTORY
	myfile.open ("history1.txt",ios::app);
	myfile<<" to "<<ants.avail_cap[hub][prod]<<endl;
	myfile.close();
#endif
    return 0;
}

int OpenHub(ind indices, data dados, ant& ants, double& temp_cost){
    
	if(ants.z[indices.hub]<1){
		ants.z[indices.hub]=1;
#ifdef HISTORY
			ofstream myfile;
			myfile.open ("history1.txt",ios::app);
			myfile<<endl<<"OPEN HUB: "<< indices.hub <<endl;
			myfile<<endl<<"update fixed costs (open hub) from: "<<endl;
			myfile << "ants.cost "<<ants.cost << endl;
			myfile << "temp_cost "<<temp_cost << endl;
			myfile.close();
#endif
        // update costs
        temp_cost = ants.cost;
        ants.cost=temp_cost+dados.g[indices.hub];
        temp_cost = ants.cost;
#ifdef HISTORY
			myfile.open ("history1.txt",ios::app);
			myfile<<endl<<" to: "<<endl;
			myfile << "ants.cost "<<ants.cost << endl;
			myfile << "temp_cost "<<temp_cost << endl;
			myfile.close();
#endif
    } //if(ants.z[indices.hub]<1)
    return 0;
} //int OpenHub(ind indices, data dados, ant& ants, double& temp_cost)

int DedicateHub(ind indices, data dados, ant& ants, aco a_param, int k, double& temp_cost){
    
    if(indices.node!=indices.hub && ants.x[indices.hub][indices.prod]!=indices.hub && ants.avail_tau[indices.hub][indices.hub][indices.prod]>0){
        
        /// por causa das funções que se chamam dentro desta
        int prod=indices.prod;
        //int node=indices.node;
        int hub=indices.hub;
        
        ants.x[hub][prod]=hub;
#ifdef HISTORY
		ofstream myfile;
		myfile.open ("history1.txt",ios::app);
		myfile<<endl<<"Solution Component Added"<<endl;
		myfile << "ants.x["<<hub<<"]["<<prod<<"]= " << hub << endl;
		myfile.close();
#endif
		// update fixed costs
#ifdef HISTORY
		myfile.open ("history1.txt",ios::app);
		myfile<<endl<<"update fixed costs (dedicate hub) from: "<<endl;
		myfile<<endl<<"update fixed costs from: "<<endl;
		myfile << "ants.cost "<<ants.cost << endl;
		myfile << "temp_cost "<<temp_cost << endl;
		myfile.close();
#endif
		temp_cost = ants.cost;
		ants.cost=temp_cost+dados.f[hub][prod];
		temp_cost = ants.cost;
#ifdef HISTORY
		myfile.open ("history1.txt",ios::app);
		myfile<<endl<<"update fixed costs to: "<<endl;
		myfile << "ants.cost "<<ants.cost << endl;
		myfile << "temp_cost "<<temp_cost << endl;
		myfile << "dados.d[hub][hub] = " << dados.d[hub][hub]<<endl;
		myfile.close();
#endif

		// Compute transfer and collection costs
		// update costs
		temp_cost = ants.cost;
		ants.cost=temp_cost+dados.d[hub][hub]*(dados.chi[prod]*dados.O[hub][prod]+dados.delta[prod]*dados.D[hub][prod]);
		temp_cost = ants.cost;
        
        // remove solution component of available solutions
        ants.avail_tau[hub][hub][prod]=0;
        
        // prevent multiple allocation of node index.hub for product p
        // is_not_hub of this function is 0 because !0=1 and this is a hub
        ApplySingleAllocationRules(prod, hub, hub, dados.nbNodes, ants, 1, 0);
        
        //Apply Lk restrictions (to only allow available solutions to be added)
        ApplyLkRules(prod, hub, dados, ants);

        // local pheromone updating
        LocalPheromoneUpdate(prod, hub, hub, a_param);
        
        // update available capacity of the opened hub
        UpdateAvailableCapacities(prod, hub, hub, dados, ants, k);
        
        /// not necessary but left here just in case
        if(ants.avail_cap[hub][prod]<=0){
			cout << "actions capacity violation "<<endl;
            system ("pause");
        }
    }
    return 0;
}

int GetBestAntCost(int nr_prods, int nr_nodes, ant& ants, iteration& iter, int it, int k){
    

    if(ants.cost<=iter.best_cost){
        iter.best_cost=ants.cost;
        iter.best_ant=k;
        
		// best solution
        // write best solution
		for(int p=0;p<nr_prods;p++)
			for(int i=0;i<nr_nodes;i++){
				iter.x_best[i][p] = ants.x[i][p];
#ifdef BAHIST
				ofstream myfile;
				myfile.open ("BESTANThistory.txt",ios::app);
				myfile<<"iter.x_best["<<i<<"]["<<p<<"] = "<<iter.x_best[i][p]<<endl;
				myfile<<"ants["<<i<<"]["<<p<<"] = "<<ants.x[i][p]<<endl;
				myfile.close();
#endif
			}
	
		for(int j=0;j<nr_nodes;j++){
			iter.z_best[j]=ants.z[j];
		}
	}

    return 0;
}

int GetBestCost(data dados, iteration iter,best& bst, int it, double& global_bst, clock_t t1){
    
    clock_t t2;
    double bst_cost=iter.best_cost;
	double aux_iter_best_cost,aux_global_bst;

	aux_iter_best_cost = bst_cost;
	aux_global_bst=global_bst;

	// BEST
    if(bst_cost<=global_bst){
        global_bst=bst_cost;
        bst.cost=bst_cost;
		bst.nr_iter=it;
		t2=clock();
		//////AAA////////
		/// global best solution
		// best solution
		// save best solution
		for(int p=0;p<dados.nbProducts;p++)
			for(int i=0;i<dados.nbNodes;i++){
				bst.x[i][p] = iter.x_best[i][p];
			}

			for(int j=0;j<dados.nbNodes;j++){
				bst.z[j]=iter.z_best[j];
			}
		//////AAA////////
        double diff ((double)t2-(double)t1);
		if(aux_iter_best_cost<aux_global_bst){
			bst.nr_iter=it;
			bst.time = diff / CLOCKS_PER_SEC;
			}
    }
    return 0;
}

int GlobalPheromoneUpdate(data dados,aco& a,iteration iter,double scl_prm){
    int p,i,j;
    double temp_tau=0.0;
	
    for(p=0;p<dados.nbProducts;p++)
        for(i=0;i<dados.nbNodes;i++){
            j=iter.x_best[i][p];
			if(j==-1)
				cout <<"j= "<<j<<endl;
            temp_tau = a.tau[i][j][p];
			a.tau[i][j][p]=(1-GAMMA)*temp_tau+scl_prm*SCALING_PARAMETER*GAMMA*(1/(iter.best_cost));
			//a.tau[j][j][p]=(1-GAMMA)*temp_tau+pow(scl_prm,SCL_P)*SCALING_PARAMETER*GAMMA*(1/(iter.best_cost));
        }
    return 0;
}

int GlobalDeadPheromoneUpdate(data dados,aco& a, ant ants, double scl_prm,double glbl_best){
    int p,i,j;
    double temp_tau=0.0;
	
    for(p=0;p<dados.nbProducts;p++)
        for(j=0;j<dados.nbNodes;j++){
#ifdef TAU_HIST
			ofstream taufile;
			taufile.open ("tau_hist.txt",ios::app);
			taufile << "tau Global updated"<< endl;
			taufile << "a.tau"<<"["<<j<<"]["<<j<<"]["<<p<<"] = "<< a.tau[j][j][p] << endl;
			taufile.close();
#endif
			temp_tau = a.tau[j][j][p];
			a.tau[j][j][p]=(1-GAMMA)*temp_tau+scl_prm*SCALING_PARAMETER*GAMMA*(1/(glbl_best/10000));
			//a.tau[j][j][p]=(1-GAMMA)*temp_tau+pow(scl_prm,SCL_P)*SCALING_PARAMETER*GAMMA*(1/(glbl_best/10000));
#ifdef TAU_HIST
			taufile.open ("tau_hist.txt",ios::app);
			taufile << "a.tau"<<"["<<j<<"]["<<j<<"]["<<p<<"] = "<< a.tau[j][j][p] << endl;
			taufile.close();
#endif
		}
    return 0;
}

File: src/com/ef/antcolony/Actions.java
package com.ef.antcolony;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.ef.antcolony.ReadData.Data;
import com.ef.antcolony.model.constants.AcoVar;
import com.ef.antcolony.utils.ArrayUtils;

public class Actions {

	private static final Logger log = LoggerFactory.getLogger(Main.class);

	// =====================================================================
	// 1. Add a solution component (x_node^prod = hub)
	// =====================================================================
	public static double addSolutionComponent(PNH indices, Data dados, Ant ant, double temp_cost) {

		// Assign node to hub for this product
		ant.x[indices.prod][indices.node] = indices.hub;

		log.error("Solution Component Added");
		log.error("ants.x[{}][{}] = {}", indices.prod, indices.node, indices.hub);

		// Remove this component from available choices
		ant.avail_tau[indices.prod][indices.node][indices.hub] = 0;

		// Update transfer + collection cost
		log.error("update transfer and collection costs from:");
		log.error("ants.cost = {}", ant.cost);
		log.error("temp_cost = {}", temp_cost);

		temp_cost = ant.cost;
		double collectionAndTransfer = dados.d[indices.node][indices.hub] *
				(dados.chi[indices.prod] * dados.originatedFlow[indices.prod][indices.node] +
						dados.delta[indices.prod] * dados.destinedFlow[indices.prod][indices.node]);

		ant.cost = temp_cost + collectionAndTransfer;
		temp_cost = ant.cost;

		log.error("update transfer and collection costs to:");
		log.error("ants.cost = {}", ant.cost);
		log.error("temp_cost = {}", temp_cost);

		// If the node is assigned to itself → it's a hub → add fixed cost per product
		if (indices.node == indices.hub) {
			log.error("update fixed costs from:");
			log.error("ants.cost = {}", ant.cost);
			log.error("temp_cost = {}", temp_cost);

			temp_cost = ant.cost;
			ant.cost = temp_cost + dados.f[indices.prod][indices.hub];
			temp_cost = ant.cost;

			log.error("update fixed costs to:");
			log.error("ants.cost = {}", ant.cost);
			log.error("temp_cost = {}", temp_cost);
		}

		return temp_cost;
	}

	// =====================================================================
	// 2. Local pheromone update (after adding a component)
	// =====================================================================
	public static void localPheromoneUpdate(int prod, int hub, int node, Aco a) {
		double oldTau = a.tau[prod][node][hub];
		log.error("tau local updated");
		log.error("a.tau[{}][{}][{}] = {}", prod, node, hub, oldTau);

		a.tau[prod][node][hub] = (1 - AcoVar.RHO) * oldTau + AcoVar.RHO * a.tau0[prod][node][hub];

		log.error("a.tau[{}][{}][{}] = {}", prod, node, hub, a.tau[prod][node][hub]);
	}

	// =====================================================================
	// 3. Apply Single Allocation Rules
	// =====================================================================
	public static void applySingleAllocationRules(int prod, int hub, int node, int nr_nodes,
			Ant ant, int is_node, int is_not_hub) {
		log.error("SOLUTION COMPONENTS MADE UNAVAILABLE SA RULES");

		// Rule 1: a node can be allocated to only one hub per product
		if (is_node > 0) {
			for (int j = 0; j < nr_nodes; j++) {
				if (ant.avail_tau[prod][node][j]> 0) {
					ant.avail_tau[prod][node][j] = 0;
					log.error("ants.avail_tau[{}][{}][{} = 0", prod, node, j);
				}
			}
		}

		// Rule 2: if a node is allocated to a hub, it cannot become a hub itself (except self-allocation)
		if (is_not_hub > 0) {
			for (int i = 0; i < nr_nodes; i++) {
				if (ant.avail_tau[prod][i][node] > 0) {
					ant.avail_tau[prod][i][node] = 0;
					log.error("ants.avail_tau[{}][{}][{}] = 0", prod, i, node);
				}
			}
		}
		
	}

	// =====================================================================
	// 4. Apply Lk (product capacity per hub) rules
	// =====================================================================
	public static void applyLkRules(int prod, int hub, Data dados, Ant ant) {
		if (dados.L[prod] >= dados.nbProducts) {
			return;  // No limit
		}

		log.error("SOLUTION COMPONENTS MADE UNAVAILABLE Lk Rules");

		HubProductCount np = HubProductCount.checkProdConnects(dados.nbProducts, dados.nbNodes, ant);

		// If this hub has reached its product capacity L[hub]
		if (np.prods[hub] >= dados.L[hub]) {
			for (int p = 0; p < dados.nbProducts; p++) {
				if (ant.x[p][hub] != hub) {  // not already a hub for product p
					for (int i = 0; i < dados.nbNodes; i++) {
						if (ant.avail_tau[p][i][hub] > 0) {
							ant.avail_tau[p][i][hub] = 0;
							log.error("ants.avail_tau[{}][{}][{}] = 0", p, i, hub);
						}
					}
				}
			}
		}
	}

	// =====================================================================
	// 5. Update available hub capacity
	// =====================================================================
	public static void updateAvailableCapacities(int prod, int hub, int node, Data dados, Ant ant, int k) {
		log.error("CAPACITY UPDATE FROM {}", ant.avail_cap[prod][hub]);
	
		double oldCap = ant.avail_cap[prod][hub];
		ant.avail_cap[prod][hub] = oldCap - dados.originatedFlow[prod][node];
		
		log.error(" to {}", ant.avail_cap[prod][hub]);
	}

	// =====================================================================
	// 6. Open a hub (pay fixed opening cost g[j])
	// =====================================================================
	public static double openHub(PNH indices, Data dados, Ant ant, double temp_cost) {

		if (ant.z[indices.hub] < 1) {
			ant.z[indices.hub] = 1;
			log.error("OPEN HUB: {}", indices.hub);
			log.error("update fixed costs (open hub) from:");
			log.error("ants.cost = {}", ant.cost);
			log.error("temp_cost = {}", temp_cost);

			temp_cost = ant.cost;
			ant.cost = temp_cost + dados.g[indices.hub];
			temp_cost = ant.cost;

			log.error(" to:");
			log.error("ants.cost = {}", ant.cost);
			log.error("temp_cost = {}", temp_cost);
		}

		return temp_cost;
	}

	// =====================================================================
	// 7. Dedicate a hub to a product (self-allocation + pay f[j][p])
	// =====================================================================
	public static double dedicateHub(PNH indices, Data dados, Ant ant, Aco a, int k, double temp_cost) {

		if (indices.node != indices.hub &&
				ant.x[indices.prod][indices.hub] != indices.hub &&
				ant.avail_tau[indices.prod][indices.hub][indices.hub] > 0) {

			int prod = indices.prod;
			int hub = indices.hub;

			ant.x[prod][hub] = hub;
			log.error("Solution Component Added (dedicate hub)");
			log.error("ants.x[{}][{}] = {}", prod, hub, hub);

			// Fixed cost for dedicating hub to product
			temp_cost = ant.cost;
			ant.cost = temp_cost + dados.f[prod][hub];

			// Transfer/collection cost for self-loop (usually 0, but kept for generality)
			temp_cost = ant.cost;
			ant.cost = temp_cost + dados.d[hub][hub] *
					(dados.chi[prod] * dados.originatedFlow[prod][hub] + dados.delta[prod] * dados.destinedFlow[prod][hub]);
			temp_cost = ant.cost;

			// Remove self-allocation from available
			ant.avail_tau[prod][hub][hub] = 0;

			// Enforce single allocation (this node is now a hub for prod)
			applySingleAllocationRules(prod, hub, hub, dados.nbNodes, ant, 1, 0);

			// Apply capacity limits
			applyLkRules(prod, hub, dados, ant);

			// Local pheromone update
			localPheromoneUpdate(prod, hub, hub, a);

			// Update capacity
			updateAvailableCapacities(prod, hub, hub, dados, ant, k);

			if (ant.avail_cap[prod][hub] <= 0) {
				log.error("actions capacity violation");
				System.exit(1);
			}
		}
		return temp_cost;
	}

	// =====================================================================
	// 8. Update iteration-best ant
	// =====================================================================
	public static void getBestAntCost(int nr_prods, int nr_nodes, Ant ant, Iteration iter, int it, int k) {
		if (ant.cost <= iter.best_cost) {
			iter.best_cost = ant.cost;
			iter.best_ant = k;

			iter.x_best = ArrayUtils.deepCopy(ant.x);
			iter.z_best = ArrayUtils.deepCopy(ant.z);
		}
	}

	// =====================================================================
	// 9. Update global best solution
	// =====================================================================
	public static int getBestCost(Data dados, Iteration iter, Best bst,
			int it, double[] globalBestHolder, long startTime) {
		double bst_cost = iter.best_cost;
		double global_bst = globalBestHolder[0];

		if (bst_cost <= global_bst) {
			global_bst = bst_cost;
			bst.cost = bst_cost;
			bst.nr_iter = it;

			long now = System.nanoTime();
			double elapsedSeconds = (now - startTime) / 1_000_000_000.0;
			bst.time = elapsedSeconds;

			// Copy solution
			bst.x = ArrayUtils.deepCopy(iter.x_best);
			bst.z = ArrayUtils.deepCopy(iter.z_best);
		}

		globalBestHolder[0] = global_bst;
		return 0;
	}

	// =====================================================================
	// 10. Global pheromone update (iteration-best reinforcement)
	// =====================================================================
	public static void globalPheromoneUpdate(Data dados, Aco a, Iteration iter, double scl_prm) {
		for (int p = 0; p < dados.nbProducts; p++) {
			for (int i = 0; i < dados.nbNodes; i++) {
				int j = iter.x_best[p][i];
				if (j == -1) {
					log.error("Error: j = -1 in global update");
					continue;
				}
				double oldTau = a.tau[p][i][j];
				a.tau[p][i][j] = (1 - AcoVar.GAMMA) * oldTau +
						scl_prm * AcoVar.SCALING_PARAMETER * AcoVar.GAMMA * (1.0 / iter.best_cost);
			}
		}
	}

	// =====================================================================
	// 11. Global pheromone update on dead hubs (encourage exploration)
	// =====================================================================
	public static void globalDeadPheromoneUpdate(Data dados, Aco a, Ant ant,
			double scl_prm, double glbl_best) {
		for (int p = 0; p < dados.nbProducts; p++) {
			for (int j = 0; j < dados.nbNodes; j++) {
				log.error("tau Global updated (dead hub)");
				log.error("a.tau[{}][{}][{}] = {}", p, j, j, a.tau[p][j][j]);

				double oldTau = a.tau[p][j][j];
				a.tau[p][j][j] = (1 - AcoVar.GAMMA) * oldTau +
						scl_prm * AcoVar.SCALING_PARAMETER * AcoVar.GAMMA * (1.0 / (glbl_best / 10000));

				log.error("a.tau[{}][{}][{}] = {}", p, j, j, a.tau[p][j][j]);
			}
		}
	}

}
